From 248f1e2d4f38d67368765ae1906935ffcfa97bef Mon Sep 17 00:00:00 2001
From: Huang Chaofan <chaofan.huang@nxp.com>
Date: Wed, 9 May 2018 17:27:41 +0800
Subject: [PATCH 3757/5242] MLK-18250 VPU: Add MU for vpu encoder and decoder
 power in dts for scfw xrdc enforcing, and add
 sync for v4l2 driver and firmware

commit  b960aa0f2fca32664d6b65aa608e9c465c583908 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add MU for vpu encoder and decoder power in dts for scfw xrdc enforcing,
and add sync for v4l2 driver and firmware

Signed-off-by: Huang Chaofan <chaofan.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm64/boot/dts/freescale/fsl-imx8qxp.dtsi |   25 ++++++++++++++++++++----
 drivers/mxc/vpu-decoder-b0/vpu_b0.c            |    4 ++--
 drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c    |    5 +++--
 3 files changed, 26 insertions(+), 8 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qxp.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8qxp.dtsi
index bfaffaf..8303704 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qxp.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qxp.dtsi
@@ -983,15 +983,32 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 
-			pd_vpu_enc: VPU_ENC {
-				reg = <SC_R_VPU_ENC_0>;
+			pd_vpu_mu_enc: VPU_ENC_MU {
+				reg = <SC_R_VPU_MU_1>;
 				#power-domain-cells = <0>;
 				power-domains =<&pd_vpu>;
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				pd_vpu_enc: VPU_ENC {
+					reg = <SC_R_VPU_ENC_0>;
+					#power-domain-cells = <0>;
+					power-domains =<&pd_vpu_mu_enc>;
+				};
 			};
-			pd_vpu_dec: VPU_DEC {
-				reg = <SC_R_VPU_DEC_0>;
+
+			pd_vpu_mu_dec: VPU_DEC_MU {
+				reg = <SC_R_VPU_MU_0>;
 				#power-domain-cells = <0>;
 				power-domains =<&pd_vpu>;
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				pd_vpu_dec: VPU_DEC {
+					reg = <SC_R_VPU_DEC_0>;
+					#power-domain-cells = <0>;
+					power-domains =<&pd_vpu_mu_dec>;
+				};
 			};
 		};
 
diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index aac356b..d61b940d 100755
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -1145,8 +1145,6 @@ static void transfer_buffer_to_firmware(struct vpu_ctx *ctx, void *input_buffer,
 
 	vpu_dbg(LVL_INFO, "enter %s, start_flag %d, index=%d, firmware_started=%d\n", __func__, ctx->start_flag, ctx->str_index, ctx->dev->firmware_started);
 
-	if (!ctx->dev->firmware_started)
-		wait_for_completion(&ctx->dev->start_cmp);
 	vpu_dbg(LVL_ALL, "firmware version is %d.%d.%d\n", (pSharedInterface->FWVersion & 0x00ff0000) >> 16, (pSharedInterface->FWVersion & 0x0000ff00) >> 8, pSharedInterface->FWVersion & 0x000000ff);
 
 	if (ctx->stream_buffer_size < buffer_size + MIN_SPACE)
@@ -2234,6 +2232,8 @@ static int v4l2_open(struct file *filp)
 			goto err_firmware_load;
 		} else
 			vpu_dbg(LVL_INFO, "done: vpu_firmware_download\n");
+		if (!ctx->dev->firmware_started)
+			wait_for_completion(&ctx->dev->start_cmp);
 		dev->fw_is_ready = true;
 	}
 	mutex_unlock(&dev->dev_mutex);
diff --git a/drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c b/drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c
index 90e2a6a..998fd84 100644
--- a/drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c
+++ b/drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c
@@ -803,8 +803,6 @@ static void v4l2_transfer_buffer_to_firmware(struct queue_data *This, struct vb2
 					__func__, ctx->start_flag, ctx->str_index,
 				ctx->dev->firmware_started);
 
-		if (!ctx->dev->firmware_started)
-			wait_for_completion(&ctx->dev->start_cmp);
 		vpu_dbg(LVL_ALL, "vpu encoder firmware version is %d.%d.%d\n",
 				(pSharedInterface->FWVersion & 0x00ff0000) >> 16,
 				(pSharedInterface->FWVersion & 0x0000ff00) >> 8,
@@ -1542,6 +1540,9 @@ static int v4l2_open(struct file *filp)
 			mutex_unlock(&dev->dev_mutex);
 			goto err_firmware_load;
 		}
+
+		if (!ctx->dev->firmware_started)
+			wait_for_completion(&ctx->dev->start_cmp);
 		dev->fw_is_ready = true;
 	}
 	mutex_unlock(&dev->dev_mutex);
-- 
1.7.9.5

